### ISSUE TITLE
Brave Leaks ISP DNS While Shields are Enabled

### ISSUE THREAD

**Issue by Balachandarsmr** [brave/brave-browser#12575]
## Description 
Brave leaks isp dns after enabling secure dns

## Steps to reproduce 

   1. Select any custom dns provider under privacy settings
   2. Go to https://browserleaks.com/dns
   3. You will see your isp dns as well as selected custom dns server 

## Actual result 
![20201108_140224](https://user-images.githubusercontent.com/29444683/98461063-379e4c00-21cf-11eb-9993-246bbd6687b8.gif)

## Expected result
Dns leak test supposed to hide isp dns while using custom dns over https server 

## Issue reproduces how often 
Easily reproduced

## Version/Channel Information:

- Can you reproduce this issue with the current Play Store version? 
No
- Can you reproduce this issue with the current Play Store Beta version? 
No
- Can you reproduce this issue with the current Play Store Nightly version?
Yes
## Device details
   - Install type (ARM, x86): ARM 32 Bit
   - Device type (Phone, Tablet, Phablet): Phone
   - Android version: 8.1
   
## Brave version
Nightly 1.18.41

### Website problems only
- Does the issue resolve itself when disabling Brave Shields? 
Yes
- Does the issue resolve itself when disabling Brave Rewards?
No
- Is the issue reproducible on the latest version of Chrome? 
No

**Comment by bsclifton** [comment]
cc: @samartnik @mkarolin

**Comment by samartnik** [comment]
it works as expected with shields off, look like we excessively block some scripts on the website

**Comment by Balachandarsmr** [comment]
Secure dns detection works with shield @ https://ipleak.net/
https://www.dnsleaktest.com/
But not on other detection sites. Yes it's due shield.

**Comment by Balachandarsmr** [comment]
This bug is not fixed yet. Without shield secure dns works. But when shield is on brave leaks isp dns.

**Comment by jonathansampson** [comment]
This issue impacts Windows users as well. Enabling the same feature in another Chromium-based browser (e.g. Chrome, Opera) shows that Brave deviates from the pack. In Chrome and Opera, https://browserleaks.com/dns lists only the DNS provider configured in the _Use Secure DNS_ browser settings. In Brave, both the user's ISP DNS and Secure DNS are listed.

**Comment by jonathansampson** [comment]
I just tested on macOS, and had the same results as those [mentioned above](https://github.com/brave/brave-browser/issues/12575#issuecomment-759850182).

**Comment by fmarier** [comment]
Seems to affect all brave-core platforms. I was able to reproduce on Linux too after patching in the DoH settings.

**Comment by Bruce-Bane** [comment]
I have the DNS leaking problem. But router DNS is leaking too

I described everything here

https://community.brave.com/t/brave-leaking-dns-servers/191002

**Comment by stephendonner** [comment]
Verified using the testplan from both https://github.com/brave/brave-core/pull/7731 and inline, here, with the (**PRE-FIX**) `1.21.50` build, compared to (**POST-FIX**) `1.20.103`, on `macOS Big Sur 11.2`

`_AT&T_` is my internet service provider ("ISP")/DNS-resolver service; Cloudflare is the DNS-over-HTTP ("DoH") provider chosen for this test.

**PRE-fix Build:**

Brave | 1.20.103 Chromium: 88.0.4324.152 (Official Build) (x86_64)
-- | --
Revision | 6579930fc53b4dc589c042bec9d0a3778326974d-refs/branch-heads/4324@{#2106}

* Verified in Wireshark there **were** UDP-based DNS-lookup leaks resolving `tools.ietf.org`
<details>
<img width="1674" alt="Screen Shot 2021-02-10 at 3 58 56 PM" src="https://user-images.githubusercontent.com/387249/107588715-3bb87c00-6bb9-11eb-925e-cb8c1a23d5ee.png">

</details>

* Verified for the test sites whether 1) just `Cloudflare` ("DoH") 2) just _`A&T`_ ("ISP") or 3) **both** _`AT&T`_ & Cloudflare ("ISP" & "DoH") were detected:
  * https://www.dnsleaktest.com (`Cloudflare`)
  * https://www.expressvpn.com/dns-leak-test (_`AT&T`_)
  * https://browserleaks.com/dns (both _`AT&T`_ & `Cloudflare`)
  * https://ipleak.net/ (`Cloudflare`)
<details>
<img width="1680" alt="BEFORE" src="https://user-images.githubusercontent.com/387249/107587576-da8fa900-6bb6-11eb-8dda-f168cda82a17.png">

</details>

**POST-fix Build:**

Brave | 1.21.50 Chromium: 88.0.4324.152 (Official Build) beta (x86_64)
-- | --
Revision | 6579930fc53b4dc589c042bec9d0a3778326974d-refs/branch-heads/4324@{#2106}
OS | macOS Version 11.2.1 (Build 20D74)
* Verified in Wireshark there were **_NO_** UDP-based DNS-lookup leaks resolving `tools.ietf.org`
<details>
<img width="1458" alt="Screen Shot 2021-02-10 at 2 35 12 PM" src="https://user-images.githubusercontent.com/387249/107587719-25a9bc00-6bb7-11eb-8613-9564ff358cb1.png">

</details>

* Verified, using the following sites, that only the DNS-over-HTTP (DoH) server was detected (that we didn't leak the `ISP` resolver -- `AT&T`, in my case):
  * https://www.dnsleaktest.com (`Cloudflare`)
  * https://www.expressvpn.com/dns-leak-test (`Cloudflare`)
  * https://browserleaks.com/dns (`Cloudflare`)
  * https://ipleak.net/ (`Cloudflare`)
 
<details>
<img width="1680" alt="AFTER" src="https://user-images.githubusercontent.com/387249/107587591-e11e2080-6bb6-11eb-80d9-a1d4bf3cbb6d.png">

</details>

**Comment by LaurenWags** [comment]
Removed `QA Pass` labels as this was uplifted to 1.20.x

**Comment by stephendonner** [comment]
Verified `FIXED` on `1.20.108` using the testplan https://github.com/brave/brave-browser/issues/12575#issuecomment-777128032 again, on:

```
Brave | 1.20.108 Chromium: 88.0.4324.182 (Official Build) (x86_64)
Revision | 73ee5087001dcef33047c4ed650471b225dd8caf-refs/branch-heads/4324@{#2202}
OS
macOS Version 11.2.1 (Build 20D74)
```

```
Brave | 1.20.108 Chromium: 88.0.4324.182 (Official Build) (64-bit)
Revision73ee5087001dcef33047c4ed650471b225dd8caf-refs/branch-heads/4324@{#2202}
OS | Windows 10 OS Version 2009 (Build 21318.1000)
```

Due to the hotfix release + time, I will update with screenshots later (promise).

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Samsung Galaxy Tablet A` running `Android 11` using the following build:

```
1.20.108 CR: 88.0.4324.182
```

**`Cloudflare (1.1.1.1)`** | **`Open DNS`** | **`Google (Public DNS)`** | **`Quad9 (9.9.9.9)`** | **`Clean Browsing (Family Filter)`**
---------------------------|------------------|------------------------------|------------------------|--------------------------------------
![Screenshot_20210219-171233_Brave](https://user-images.githubusercontent.com/2602313/108569199-787f2400-72d9-11eb-93ce-872b1da2b982.jpg) | ![Screenshot_20210219-171528_Brave](https://user-images.githubusercontent.com/2602313/108569328-b4b28480-72d9-11eb-9b70-a54ac670c5ea.jpg) | ![Screenshot_20210219-171629_Brave](https://user-images.githubusercontent.com/2602313/108569399-d3188000-72d9-11eb-8165-b06fee6d563f.jpg) | ![Screenshot_20210219-171802_Brave](https://user-images.githubusercontent.com/2602313/108569646-3b676180-72da-11eb-89ad-42695b5a8cad.jpg) | ![Screenshot_20210219-173213_Brave](https://user-images.githubusercontent.com/2602313/108569715-5fc33e00-72da-11eb-9765-79085348791e.jpg)
